scripts/evaluate_heartbeat.py

```diff
@@ -294,65 +280,134 @@ def investigate_offenders(
+    try:
+        dkg_timeout_secs = coordinator.dkgTimeout()
+        dkg_timeout = timedelta(seconds=dkg_timeout_secs)
+    except Exception:
+        click.secho("⚠️ Failed to fetch DKG timeout from coordinator", fg="red")
+        return
 
     offenders: Dict[str, Dict[str, Any]] = defaultdict(dict)
 
     if heartbeat_round > 4:
         click.secho(
-            f"⚠️ This is the heartbeat round #{heartbeat_round}, which exceeds the expected maximum"
-            + " of 4 per month.",
+            f"⚠️ This is the heartbeat round #{heartbeat_round}, which exceeds"
+            + " the expected maximum of 4 per month.",
             fg="yellow",
         )
-        click.secho("Skipping heartbeat evaluation...")
+        click.secho("Skipping heartbeat evaluation...", fg="yellow")
         return
 
     network_data = get_taco_network_data(domain)
 
-    for ritual_id, cohort in artifact_data.items():
-        ritual_status = coordinator.getRitualState(ritual_id)
-
-        if ritual_status == RitualState.ACTIVE.value:
-            continue  # Skip active rituals
-
-        if ritual_status == RitualState.DKG_TIMEOUT.value:
+    for ritual_id, _ in artifact_data.items():
+        try:
+            ritual_status = coordinator.getRitualState(ritual_id)
             participants = coordinator.getParticipants(ritual_id)
-
-            for participant_info in participants:
-                address, aggregated, transcript, *data = participant_info
-
-                if not transcript:
-                    offenders[ritual_id][address] = {"reasons": [MISSING_TRANSCRIPT]}
-
+            init_timeout_timestamp, _ = coordinator.getTimestamps(ritual_id)
+        except Exception as e:
+            click.secho(f"⚠️ Failed to fetch ritual data for {ritual_id}: {e}", fg="red")
+            return
+
+        # let's check if we gave enough time to timeout: no rituals in progress
+        init_timeout = datetime.fromtimestamp(init_timeout_timestamp, tz=timezone.utc)
+        if init_timeout + dkg_timeout > datetime.now(tz=timezone.utc):
+            click.secho(
+                f"⚠️ The DKG ritual {ritual_id} is still within the timeout period."
+                + " The evaluation was run too early?",
+                fg="red",
+            )
+            return
+
+        for participant_info in participants:
+            address, _, transcript, _ = participant_info
+
+            version = get_node_version(address, network_data)
+
+            offenders[address] = {"ritual": ritual_id, "reasons": [], "version": version}
+
+            # check if node is reachable and running a valid version
+            if version == UNREACHABLE:
+                offenders[address]["reasons"].append(UNREACHABLE)
+                click.secho(f"Node {address} is unreachable", fg="cyan")
+            else:
+                try:
+                    version = Version(version)
+                    if version not in valid_versions:
```

Comment:
would this include new version that are not part of any release yet? like experimental? or we don't really care about them?

Reply:
This would only be for published releases - https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#list-releases

Reply:
Oh I see what you mean. Like someone running an unreleased (but newer eg. v7.7.0) experimental build on mainnet when a heartbeat round occurs.

@manumonti maybe you could also do a check if the version is > latest release, and if so then that be ok...?

Reply:
I'm addressing this in a following PR